File: demos/demos/common/utils/permutation.py
# File: permutation.py

from __future__ import division, unicode_literals

import math


def permute(iterable, index):
    
    seq = list(iterable)
    fact = math.factorial(len(seq))
    index %= fact
    perm = []
    
    while seq:
        fact //= len(seq)
        next, index = divmod(index, fact)
        item = seq.pop(next)
        perm.append(item)
    
    return perm


def permute_ori(iterable, index):
    
    seq = list(iterable)
    seqlen = len(seq)
    
    fact = math.factorial(seqlen)
    index %= fact
    
    next = []
    
    for i in range(seqlen, 0, -1):
        fact //= i
        pos, index = divmod(index, fact)
        next.append(pos)
    
    perm = []
    
    for item, pos in zip(reversed(seq), reversed(next)):
        perm.insert(pos, item)
    
    return perm

File: demos_voting/base/static/base/permutation.js
// Factorial, n-th permutation and original from n-th permutation functions.
// A sjcl version with division support is required.

function factorial(n) {

    var n = new sjcl.bn(n);
    var i = new sjcl.bn(2);
    var val = new sjcl.bn(1);

    while(n.greaterEquals(i)) {
        val = val.mul(i);
        i.addM(1);
    }

    return val;
}

function permute(iterable, index) {

    var seq = iterable.slice();
    var fact = factorial(seq.length);
    var perm = new Array();
    var next, index = (new sjcl.bn(index)).mod(fact);

    while (seq.length > 0) {

        fact = fact.div(seq.length);
        next = index.divmod(fact, index);
        item = seq.splice(parseInt(next.toString(), 16), 1);
        perm.push(item);
    }

    return perm;
}

function permute_ori(iterable, index) {

    var seq = iterable.slice();

    var fact = factorial(seq.length);
    var pos, index = (new sjcl.bn(index)).mod(fact);
    var next = new Array();

    for (var i = seq.length; i > 0; i--) {

        fact = fact.div(i);
        pos = index.divmod(fact, index);
        next.push(pos);
    }

    var pos, item;
    var perm = new Array();

    for (var i = seq.length - 1; i >= 0; i--) {
        pos = next[i];
        item = seq[i];
        perm.splice(pos, 0, item)
    }

    return perm;
}

